### PR TITLE
Add paddle backend

### PR DESCRIPTION
**Pull request by HydrogenSulfate**
## Category

- [x] New feature
- [ ] Bugfix
- [ ] Breaking change
- [ ] Refactoring
- [x] Documentation
- [ ] Other (please explain)

## Description

As talked in <https://github.com/NVIDIA/warp/discussions/285>, we add PaddlePaddle framework as one of backends of warp.

Thanks for reply from @shi-eric(<https://github.com/NVIDIA/warp/pull/313#issuecomment-2380075250>), I create this PR as contained only one commit.

This PR additionally adds support for the bool type to warp’s dlpack, based on this discussion: <https://github.com/dmlc/dlpack/issues/75> and the corresponding PR: <https://github.com/dmlc/dlpack/pull/76>. The boolean type representation seems uses 8 bits instead of the current 1 bit finally. I am not sure if there were any additional considerations regarding this in warp’s development. If there are specific explanations, please let me know. I will adjust the support for the boolean type of paddle, according to the suggestions, thanks :).

run test:

``` sh
# 1. install paddlepaddle
# cu123
$ python -m pip install --pre paddlepaddle-gpu -i https://www.paddlepaddle.org.cn/packages/nightly/cu123/
# cu118
# python -m pip install --pre paddlepaddle-gpu -i https://www.paddlepaddle.org.cn/packages/nightly/cu118/

# 2. install warp
$ cd warp/
$ python build.py
$ pip install -e .

# 3. run test related to paddle framework
$ python warp/tests/test_dlpack.py -k paddle
Warp 1.3.3 initialized:
   CUDA Toolkit 11.6, Driver 12.0
   Devices:
     "cpu"      : "x86_64"
     "cuda:0"   : "Tesla V100-PCIE-16GB" (16 GiB, sm_70, mempool enabled)
   Kernel cache:
     /root/.cache/warp/1.3.3
Skipping Jax DLPack tests due to exception: No module named 'jax'
W0928 17:58:46.172823 11252 gpu_resources.cc:119] Please NOTE: device: 0, GPU Compute Capability: 7.0, Driver API Version: 12.0, Runtime API Version: 11.6
W0928 17:58:46.174096 11252 gpu_resources.cc:164] device: 0, cuDNN Version: 8.4.
Skipping Paddle DLPack tests due to exception: name 'test_dlpack_paddle_to_warp_v2' is not defined
test_dlpack_paddle_to_warp_cuda_0 (__main__.TestDLPack) ... Module __main__ cac4f71 load on device 'cuda:0' took 373.07 ms  (compiled)
ok
test_dlpack_warp_to_paddle_cuda_0 (__main__.TestDLPack) ... ok
test_dlpack_warp_to_paddle_v2_cuda_0 (__main__.TestDLPack) ... ok

----------------------------------------------------------------------
Ran 3 tests in 0.498s

OK

$ python warp/tests/test_dlpack.py -k paddle
Warp 1.3.3 initialized:
   CUDA Toolkit 11.6, Driver 12.0
   Devices:
     "cpu"      : "x86_64"
     "cuda:0"   : "Tesla V100-PCIE-16GB" (16 GiB, sm_70, mempool enabled)
   Kernel cache:
     /root/.cache/warp/1.3.3
Skipping Jax DLPack tests due to exception: No module named 'jax'
W0928 13:15:33.230201 29239 gpu_resources.cc:119] Please NOTE: device: 0, GPU Compute Capability: 7.0, Driver API Version: 12.0, Runtime API Version: 11.6
W0928 13:15:33.231549 29239 gpu_resources.cc:164] device: 0, cuDNN Version: 8.4.
test_dlpack_paddle_to_warp_cuda_0 (__main__.TestDLPack) ... Module __main__ cac4f71 load on device 'cuda:0' took 387.06 ms  (compiled)
ok
test_dlpack_paddle_to_warp_v2_cuda_0 (__main__.TestDLPack) ... ok
test_dlpack_warp_to_paddle_cuda_0 (__main__.TestDLPack) ... ok
test_dlpack_warp_to_paddle_v2_cuda_0 (__main__.TestDLPack) ... ok

----------------------------------------------------------------------
Ran 4 tests in 0.567s

OK
```

todolist:

- [x] Add `__cuda_array_interface__` in paddlepaddle(<https://github.com/PaddlePaddle/Paddle/pull/68192>)
- [x] Merge pr for supporting zero-copy dlpack(<https://github.com/PaddlePaddle/Paddle/pull/67927>)
## Changelog

- Add specific line-by-line info of high level changes in this PR.

## Before your PR is "Ready for review"

- [x] Do you agree to the terms under which contributions are accepted as described in Section 9 the Warp [License](https://github.com/NVIDIA/warp/blob/main/LICENSE.md)?
- [x] Have you read the [Contributor Guidelines](https://github.com/NVIDIA/warp/blob/main/CONTRIBUTING.md)?
- [x] Have you written any new necessary tests?
- [x] Have you added or updated any necessary documentation?
- [x] Have you added any files modified by compiling Warp and building the documentation to this PR (.e.g. `stubs.py`, `functions.rst`)?
- [x] Does your code pass `ruff check` and `ruff format --check`?

### CODE REVIEW

**Comment by shi-eric** [comment]
Did you forget to include `paddle.py`?

File: docs/installation.rst [hunk 1]
```diff
@@ -76,6 +76,7 @@ The following optional dependencies are required to support certain features:
 * `usd-core <https://pypi.org/project/usd-core>`_: Required for some Warp examples, ``warp.sim.parse_usd()``, and ``warp.render.UsdRenderer``.
 * `JAX <https://jax.readthedocs.io/en/latest/installation.html>`_: Required for JAX interoperability (see :ref:`jax-interop`).
 * `PyTorch <https://pytorch.org/get-started/locally/>`_: Required for PyTorch interoperability (see :ref:`pytorch-interop`).
+* `Paddle <https://www.paddlepaddle.org.cn/>`_: Required for Paddle interoperability (see :ref:`paddle-interop`).
```
Comment (shi-eric) [inline docs/installation.rst]: Maybe link to the GitHub instead? https://github.com/PaddlePaddle/Paddle In the US, the website doesn't seem to load.

Comment (HydrogenSulfate) [reply docs/installation.rst]: > Maybe link to the GitHub instead? https://github.com/PaddlePaddle/Paddle In the US, the website doesn't seem to load.

I did forget to upload `paddle.py` and `test_paddle.py`😥. It has now been uploaded along with the changes to the paddle link through `git commit --amend`.

**Comment by mmacklin** [comment]
Thank you for the very nice PR! This is really well done, we are aiming to merge this with our 1.4.0 release this week.

**Comment by HydrogenSulfate** [comment]
> Thank you for the very nice PR! This is really well done, we are aiming to merge this with our 1.4.0 release this week.

Thanks a lot to @mmacklin and @shi-eric for your review. Note that there seems still a link error in `build-and-test / pull-request-docs` CI, which might because 
 the reason @shi-eric mentioned(<https://github.com/NVIDIA/warp/pull/318#discussion_r1779731457>), I am afraid of warp's document CI will be affected by this. 
image
Please check if `https://www.paddlepaddle.org.cn/` be accessed in US, or please help to replace it with `https://github.com/PaddlePaddle/Paddle`

**Comment by shi-eric** [comment]
> > Thank you for the very nice PR! This is really well done, we are aiming to merge this with our 1.4.0 release this week.
> 
> Thanks a lot to @mmacklin and @shi-eric for your review. Note that there seems still a link error in `build-and-test / pull-request-docs` CI, which might because the reason @shi-eric mentioned([#318 (comment)](https://github.com/NVIDIA/warp/pull/318#discussion_r1779731457)), I am afraid of warp's document CI will be affected by this. image Please check if `https://www.paddlepaddle.org.cn/` be accessed in US, or please help to replace it with `https://github.com/PaddlePaddle/Paddle`

Yeah, I fixed that with ab147a089c70c4f5bdf97024714b6ddef28d775d so we're all good. I did notice that when I tried to install Paddle (CUDA 12.3) and run `test_paddle.py` on my (multi-GPU) system, I got a bunch of CUDA-related errors. I'll let you know if I have a chance to look into it...

**Comment by HydrogenSulfate** [comment]
> > > Thank you for the very nice PR! This is really well done, we are aiming to merge this with our 1.4.0 release this week.
> > 
> > 
> > Thanks a lot to @mmacklin and @shi-eric for your review. Note that there seems still a link error in `build-and-test / pull-request-docs` CI, which might because the reason @shi-eric mentioned([#318 (comment)](https://github.com/NVIDIA/warp/pull/318#discussion_r1779731457)), I am afraid of warp's document CI will be affected by this. image Please check if `https://www.paddlepaddle.org.cn/` be accessed in US, or please help to replace it with `https://github.com/PaddlePaddle/Paddle`
> 
> Yeah, I fixed that with [ab147a0](https://github.com/NVIDIA/warp/commit/ab147a089c70c4f5bdf97024714b6ddef28d775d) so we're all good. I did notice that when I tried to install Paddle (CUDA 12.3) and run `test_paddle.py` on my (multi-GPU) system, I got a bunch of CUDA-related errors. I'll let you know if I have a chance to look into it...

Thanks for fixing that, I found it may because not setting `device_id` correctly in `FromDLpack` of paddle
 framework, so when a tensor from gpu ohter than the first device(id=0), paddle still setting the default id(0) for converted tensor(https://github.com/PaddlePaddle/Paddle/blob/e549437ad8bda6ff55ed22039eb09bafa74f78ba/paddle/fluid/framework/tensor_util.cc#L910-L911). I will verify and fix it soonly.

**Pull request merged**